modules/javafx.graphics/src/main/java/com/sun/glass/ui/GlassRobot.java

```diff
@@ -0,0 +1,341 @@
+     * @param y the starting y-position of the rectangular area to capture
+     * @param width the width of the rectangular area to capture
+     * @param height the height of the rectangular area to capture
+     * @param data the array to fill with the raw pixel data corresponding to
+     * the captured region
+     */
+    public abstract void getScreenCapture(int x, int y, int width, int height, int[] data);
+
+    /**
+     * Returns an {@code Image} containing the specified rectangular area of the screen.
+     * <p>
+     * If the {@code scaleToFit} argument is {@literal false}, the returned
+     * {@code Image} object dimensions may differ from the requested {@code width}
+     * and {@code height} depending on how many physical pixels the area occupies
+     * on the screen. E.g. in HiDPI mode on the Mac (aka Retina display) the pixels
+     * are doubled, and thus a screen capture of an area of size (10x10) pixels
+     * will result in an {@code Image} with dimensions (20x20). Calling code should
+     * use the returned images's {@link Image#getWidth() and {@link Image#getHeight()
+     * methods to determine the actual image size.
+     * <p>
+     * If {@code scaleToFit} is {@literal true}, the returned {@code Image} is of
+     * the requested size. Note that in this case the image will be scaled in
+     * order to fit to the requested dimensions if necessary such as when running
+     * on a HiDPI display.
+     *
+     * @param x the starting x-position of the rectangular area to capture
+     * @param y the starting y-position of the rectangular area to capture
+     * @param width the width of the rectangular area to capture
+     * @param height the height of the rectangular area to capture
+     * @param scaleToFit If {@literal true} the returned {@code Image} will be
+     * scaled to fit the request dimensions, if necessary. Otherwise the size
+     * of the returned image will depend on the output scale (DPI) of the primary
+     * screen.
+     */
+    public final WritableImage getScreenCapture(WritableImage image, double x, double y, double width,
+                                                double height, boolean scaleToFit) {
+        javafx.stage.Screen primaryScreen = Screen.getPrimary();
+        double outputScaleX = primaryScreen.getOutputScaleX();
+        double outputScaleY = primaryScreen.getOutputScaleY();
+        int data[];
+        int dw, dh;
+        if (outputScaleX == 1.0f && outputScaleY == 1.0f) {
+            // No scaling with be necessary regardless of if "scaleToFit" is set or not.
+            data = new int[(int) (width * height)];
+            getScreenCapture((int) x, (int) y, (int) width, (int) height, data);
+            dw = (int) width;
+            dh = (int) height;
+        } else {
+            // Compute the absolute pixel bounds that the requested size will fill given
+            // the display's scale.
+            int pminx = (int) Math.floor(x * outputScaleX);
```

Comment:
most variables can be defined as `final`

Reply:
Besides aiding comprehension for developer's reading/changing this code in the future, is there any additional benefit?